cpp/oneapi/dal/backend/primitives/optimizers/common.hpp

```diff
@@ -0,0 +1,149 @@
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#pragma once
+
+#include "oneapi/dal/backend/primitives/ndarray.hpp"
+#include "oneapi/dal/backend/primitives/objective_function/logloss.hpp"
+#include "oneapi/dal/backend/primitives/blas/gemv.hpp"
+#include "oneapi/dal/backend/primitives/element_wise.hpp"
+
+namespace oneapi::dal::backend::primitives {
+
+template <typename Float>
+sycl::event l1_norm(sycl::queue& queue,
+                    const ndview<Float, 1>& x,
+                    Float* res_gpu,
+                    Float* res_host,
+                    const event_vector& deps = {});
+
+template <typename Float>
+sycl::event dot_product(sycl::queue& queue,
+                        const ndview<Float, 1>& x,
+                        const ndview<Float, 1>& y,
+                        Float* res_gpu,
+                        Float* res_host,
+                        const event_vector& deps = {});
+
+template <typename Float>
+class BaseMatrixOperator {
+public:
+    virtual ~BaseMatrixOperator() {}
+    virtual sycl::event operator()(const ndview<Float, 1>& vec,
+                                   ndview<Float, 1>& out,
+                                   const event_vector& deps = {}) = 0;
+};
+
+template <typename Float>
+class LinearMatrixOperator : public BaseMatrixOperator<Float> {
+public:
+    LinearMatrixOperator(sycl::queue& q, const ndview<Float, 2>& A);
+    ~LinearMatrixOperator() {}
+    sycl::event operator()(const ndview<Float, 1>& vec,
+                           ndview<Float, 1>& out,
+                           const event_vector& deps) override {
+        ONEDAL_ASSERT(A_.get_dimension(1) == vec.get_dimension(0));
+        ONEDAL_ASSERT(out.get_dimension(0) == vec.get_dimension(0));
+        sycl::event fill_out_event = fill<Float>(q_, out, Float(0), deps);
+        return gemv(q_, A_, vec, out, Float(1), Float(0), { fill_out_event });
+    }
+
+private:
+    sycl::queue q_;
+    const ndview<Float, 2> A_;
+};
+
+template <typename Float>
+class BaseFunction {
+public:
+    virtual ~BaseFunction() {}
+    virtual Float get_value() = 0;
+    virtual ndview<Float, 1>& get_gradient() = 0;
+    virtual BaseMatrixOperator<Float>& get_hessian_product() = 0;
+    virtual sycl::event update_x(const ndview<Float, 1>& x,
+                                 bool needHessp = false,
+                                 const event_vector& deps = {}) = 0;
+};
+
+// f(x) = 1/2 x^t A x - b^t x
+// df / dx = Ax - b
+// df / d^2x = A
+template <typename Float>
```

Comment:
I believe this function is needed only for testing purposes. I insist on moving it into the test compiled files